### DISCUSSION TITLE
Invalid operation since v3.0.2: NNN has not been declared in the entity data model

### DISCUSSION THREAD

**Discussion by dettner** [AutoMapper/AutoMapper.Extensions.OData#151] [Q&A] [answered]
_I am sorry if this is the wrong forum, please redirect me to the correct place, if that is the case._

With AutoMapper.AspNetCore.OData.EFCore 3.0.1, everything works(tm).

When upgrading to v3.0.2 or later I get the exception below, whenever I add _any_ OData query to my OData resource (e.g. a simple  "/?$top=1")

`System.InvalidOperationException: The type 'CorrectFull.NameOfMyEDMType' has not been declared in the entity data model.
   at AutoMapper.AspNet.OData.ODataQueryContextExtentions.FindSortableProperties(ODataQueryContext context, Type type)
   at AutoMapper.AspNet.OData.LinqExtensions.GetQueryableMethod(Expression expression, ODataQueryContext context, OrderByClause orderByClause, Type type, Nullable'1 skip, Nullable'1 top)
   at AutoMapper.AspNet.OData.LinqExtensions.GetOrderByMethod[T](Expression expression, ODataQueryOptions'1 options, ODataSettings oDataSettings)
   at AutoMapper.AspNet.OData.LinqExtensions.GetQueryableExpression[T](ODataQueryOptions'1 options, ODataSettings oDataSettings)
   at AutoMapper.AspNet.OData.QueryableExtensions.GetQueryable[TModel,TData](IQueryable'1 query, IMapper mapper, ODataQueryOptions'1 options, QuerySettings querySettings, Expression'1 filter)` ...

There are no other changes on my side except for the v3.0.1 -> v3.0.2 version upgrade. I have tried AutoMapper.AspNetCore.OData.EFCore v3.0.5 with the same result.

I use .NET6, and I have the latest versions of everything else.

The OData resource is defined like this:

`var myBuilder = builder.Services.AddControllersWithViews()
    .AddOData(options =>
    {
        options.AddRouteComponents("odataapi", Startup.BuildEdmModelForOData()).EnableQueryFeatures();
    });`

that static method is defined like this:
`public static IEdmModel BuildEdmModelForOData()
        {
            var builder = new ODataConventionModelBuilder { Namespace = "CorrectFull" };
            builder.EntitySet<NameOfMyEDMType>("NameOfMyEDMTypes");
            var result = builder.GetEdmModel();
            return result;
        }`

I would really appreciate some help.

**Comment by BlaiseD** [comment]
The error is being thrown [here](https://github.com/AutoMapper/AutoMapper.Extensions.OData/blob/8b5d4896adc9404ea7ec581826b3327665062b15/AutoMapper.AspNetCore.OData.EFCore/ODataQueryContextExtentions.cs#L17) it could be a bug.  The change was added in [PR 123](https://github.com/AutoMapper/AutoMapper.Extensions.OData/pull/123).

Since our tests and samples all work I would compare your configuration to the [samples](https://github.com/AutoMapper/AutoMapper.Extensions.OData/blob/8b5d4896adc9404ea7ec581826b3327665062b15/WebAPI.OData.EFCore/Startup.cs#L34) and recommend a fix if some valid configuration is not being supported.

**Reply by dettner** [reply]
Thank you @BlaiseD , I will do as you would do.

**Comment by dettner** [comment]
The title of PR 123 (_Adds support for $skip and $top with no order specified_) that you kindly pointed me to, led me to try **adding an "orderby"** to my "top" query.

Then the query works even in v3.0.5 without the exception!

Perhaps my configuration somehow provokes it. 

It seems PR 123 assumes something that my code or configuration does not live up to.

 I can't see anything obvious but I will keep on looking.

**Comment by BlaiseD** [comment, marked as answer]
I'll convert this to an issue.

Looks like you have a custom namespace  which means [here](https://github.com/AutoMapper/AutoMapper.Extensions.OData/blob/8b5d4896adc9404ea7ec581826b3327665062b15/AutoMapper.AspNetCore.OData.EFCore/ODataQueryContextExtentions.cs#L14) it can't find an `IEDMEnityType` from the type full name.

Similar to issue #149.

**Comment by BlaiseD** [comment]
Adding `$orderby` works around this logic.  Thanks for reporting.

**Comment by dettner** [comment]
You are completely correct. Confirming that

`var builder = new ODataConventionModelBuilder { Namespace = "MyNameSpace" };`

provokes the error when doing "top" without "orderby" in v3.0.2 and later versions. However:

`var builder = new ODataConventionModelBuilder(); // note no namespace`

**works** in 3.0.2 and later. 

I am happy! Thank you @BlaiseD
